qoqo/build.rs

```diff
@@ -161,6 +168,184 @@ const SOURCE_FILES: &[&str] = &[
+    if module == "qoqo" {
+        module_doc.push_str("from typing import Optional, List, Tuple, Dict, Set  # noqa: F401\nfrom qoqo_quest import Backend\n\n");
+    } else {
+        module_doc.push_str("from .qoqo import Circuit, Operation  # noqa: F401\nimport numpy as np  # noqa: F401\nfrom typing import Tuple, List, Optional, Set, Dict, Union, Self, Sequence  # noqa: F401\n\n");
+    };
+    pyo3::prepare_freethreaded_python();
+    Python::with_gil(|py| -> PyResult<String> {
+        let python_module = PyModule::import_bound(py, module)?;
+        let dict = python_module.as_gil_ref().getattr("__dict__")?;
+        let r_dict = dict.downcast::<PyDict>()?;
+        for (fn_name, func) in r_dict.iter() {
+            let name = fn_name.str()?.extract::<String>()?;
+            if name.starts_with("__")
+                || (module == "qoqo"
+                    && ![
+                        "qoqo",
+                        "Circuit",
+                        "QuantumProgram",
+                        "CircuitDag",
+                        "operations",
+                    ]
+                    .contains(&name.as_str()))
+            {
+                continue;
+            }
+            let doc = func.getattr("__doc__")?.extract::<String>()?;
+            if name == "qoqo" {
+                module_doc.push_str(&format!(
+                    "def {name}({}):\n    \"\"\"\n{doc}\n\"\"\"\n\n",
+                    collect_args_from_doc(doc.as_str()).join(", "),
+                ));
+            } else if name == "operations" {
+                module_doc.push_str(&format!(
+                    "class Operation:\n    \"\"\"\n{doc}\n\"\"\"\n\n    def __init__(self):\n       return\n\n",
+                ));
+            } else {
+                let args = collect_args_from_doc(doc.as_str()).join(", ");
+                module_doc.push_str(&format!(
+                    "class {name}{}:\n    \"\"\"\n{doc}\n\"\"\"\n\n    def __init__(self{}):\n       return\n\n",
+                    module.eq("qoqo.operations").then(|| "(Operation)").unwrap_or_default(),
+                    if args.is_empty() { "".to_owned() } else { format!(", {}", args) },
+                ));
+                let class_dict = func.getattr("__dict__")?;
+                let items = class_dict.call_method0("items")?;
+                let dict_obj = py
+                    .import_bound("builtins")?
+                    .call_method1("dict", (items,))?;
+                let class_r_dict = dict_obj.as_gil_ref().downcast::<PyDict>()?;
+                for (class_fn_name, meth) in class_r_dict.iter() {
+                    let meth_name = class_fn_name.str()?.extract::<String>()?;
+                    let class_doc = match meth_name.as_str() {
+                        "__add__" => r#"Implement the `+` (__add__) magic method to add two Circuits.
```

Comment:
How come `__add__` and `__new__` are the only functions here?

Reply:
Also, should this contain references to Circuit?

Reply:
This part is to override the doc given by __doc__ for certain methods.
 I have __add__ and __iadd__ because with the __doc__ method we get the default python documentation, but we need this one so that when you do circuit += ops.Hadamard(0) the += returns a Circuit otherwhise the circuit variable will then be considered as an Any and we loose all the hints. I probably should add an `if name.eq(&"Circuit") in these two matches because I only want them for Circuits (luckly it seems only circuits implements it so it worked)
 I also have __new__ for every classes because otherwhise it will give the default python methods and show it instead of what I put in the __init__ as the int for a class